iceoryx_hoofs/source/cli/command_line_argument_parser.cpp

```diff
@@ -0,0 +1,410 @@
+    {
+        if (option.isSameOption(value))
+        {
+            isOptionSet = true;
+            break;
+        }
+    }
+
+    if (isOptionSet)
+    {
+        std::cout << "The option \"" << value << "\" is already set!" << std::endl;
+        printHelpAndExit();
+    }
+
+    return isOptionSet;
+}
+
+bool CommandLineArgumentParser::doesOptionValueFitIntoString(const char* value) const noexcept
+{
+    const bool doesOptionValueFitIntoString = doesFitIntoString(value, MAX_OPTION_ARGUMENT_LENGTH);
+
+    if (!doesOptionValueFitIntoString)
+    {
+        std::cout << "\"" << value << "\" is longer then the maximum supported size of " << MAX_OPTION_ARGUMENT_LENGTH
+                  << " for option values." << std::endl;
+        printHelpAndExit();
+    }
+
+    return doesOptionValueFitIntoString;
+}
+
+bool CommandLineArgumentParser::hasLexicallyValidOption(const char* value) const noexcept
+{
+    return doesOptionStartWithDash(value) && hasNonEmptyOptionName(value) && hasValidShortOptionDashCount(value)
+           && hasValidOptionDashCount(value) && doesOptionNameFitIntoString(value);
+}
+
+void CommandLineOptionSet::sortAvailableOptions() noexcept
+{
+    std::sort(m_availableOptions.begin(), m_availableOptions.end());
+}
+
+CommandLineOptionValue CommandLineArgumentParser::parse(const CommandLineOptionSet& optionSet,
+                                                        int argc,
+                                                        char* argv[],
+                                                        const uint64_t argcOffset,
+                                                        const UnknownOption actionWhenOptionUnknown) noexcept
+{
+    m_optionSet = &optionSet;
+
+    m_argc = static_cast<uint64_t>(algorithm::max(0, argc));
+    m_argv = argv;
+    m_argcOffset = argcOffset;
+    // reset options otherwise multiple parse calls work on already parsed options
+    m_optionValue = CommandLineOptionValue();
+
+    if (!hasArguments(m_argc) || !assignBinaryName(m_argv[0]))
+    {
+        return m_optionValue;
+    }
+
+    for (uint64_t i = algorithm::max(argcOffset, static_cast<uint64_t>(1U)); i < m_argc; ++i)
+    {
+        const auto skipCommandLineArgument = [&] { ++i; };
+
+        if (!hasLexicallyValidOption(m_argv[i]))
+        {
+            return m_optionValue;
+        }
+
+        uint64_t optionNameStart = (m_argv[i][1] == '-') ? 2 : 1;
+        auto optionEntry = m_optionSet->getOption(OptionName_t(cxx::TruncateToCapacity, m_argv[i] + optionNameStart));
+
+        if (!optionEntry)
+        {
+            switch (actionWhenOptionUnknown)
+            {
+            case UnknownOption::TERMINATE:
+            {
+                std::cout << "Unknown option \"" << m_argv[i] << "\"" << std::endl;
+                printHelpAndExit();
+                return m_optionValue;
+            }
+            case UnknownOption::IGNORE:
+            {
+                if (isNextArgumentAValue(i))
```

Comment:
Pleas print a message that an unknown option was found and that it will be ignored.

Reply:
I explicitly do not want that. Let's assume you do something like ros with the commands and subcommands and parse some arguments internally and the remaining arguments are forwarded to another command line parser which knows what to do with it.

Then it would be absolutely annoying to see messages on the console that some arguments weren't parsed by the first parser.

Reply:
But what ros does with the subcommands is different that what this does, doesn't it?

Let's assume I have `./foo bar baz caramba`. Then I would assume that the parser would stop parsing when the subcommand `bar` is found and delegate the remaining arguments to the next stage. If there is no output, how do I know that I mistyped the command like `./foo baz baz caramba`?

Reply:
... when thinking about it, there shouldn't be an option to ignore params but just the option to stop parsing when a subcommand is found or maybe call the nesting parser on the remaining arguments

Reply:
I removed the feature again. Whenever you type an unknown option you get an error message and the help is shown.